data_types/src/partition_template.rs

```diff
@@ -1,81 +1,272 @@
-        namespace: Option<&'a Arc<NamespacePartitionTemplateOverride>>,
-        default: &'a DefaultPartitionTemplate,
-    ) -> &'a PartitionTemplate {
-        table
-            .map(|t| &t.0)
-            .or(namespace.map(|n| &n.0))
-            .unwrap_or(default.0)
+/// When a table is being created implicitly by a write, there is no possibility of a user-supplied
+/// partition template, so the table will get the namespace's partition template.
+impl From<&NamespacePartitionTemplateOverride> for TablePartitionTemplateOverride {
+    fn from(namespace_template: &NamespacePartitionTemplateOverride) -> Self {
+        Self(Arc::clone(&namespace_template.0))
     }
 }
 
-/// `TemplatePart` specifies what part of a row should be used to compute this
-/// part of a partition key.
-#[derive(Debug, Eq, PartialEq, Clone)]
-pub enum TemplatePart {
-    /// The value in a named column
-    Column(String),
-    /// Applies a  `strftime` format to the "time" column.
-    ///
-    /// For example, a time format of "%Y-%m-%d %H:%M:%S" will produce
-    /// partition key parts such as "2021-03-14 12:25:21" and
-    /// "2021-04-14 12:24:21"
-    TimeFormat(String),
+impl<DB> sqlx::Type<DB> for TablePartitionTemplateOverride
+where
+    sqlx::types::Json<Self>: sqlx::Type<DB>,
+    DB: sqlx::Database,
+{
+    fn type_info() -> DB::TypeInfo {
+        <sqlx::types::Json<Self> as sqlx::Type<DB>>::type_info()
+    }
+}
+
+impl<'q, DB> sqlx::Encode<'q, DB> for TablePartitionTemplateOverride
+where
+    DB: sqlx::Database,
+    for<'b> sqlx::types::Json<&'b proto::PartitionTemplate>: sqlx::Encode<'q, DB>,
+{
+    fn encode_by_ref(
+        &self,
+        buf: &mut <DB as sqlx::database::HasArguments<'q>>::ArgumentBuffer,
+    ) -> sqlx::encode::IsNull {
+        <sqlx::types::Json<&proto::PartitionTemplate> as sqlx::Encode<'_, DB>>::encode_by_ref(
+            &sqlx::types::Json(&self.0),
+            buf,
+        )
+    }
+}
+
+impl<'q, DB> sqlx::Decode<'q, DB> for TablePartitionTemplateOverride
+where
+    DB: sqlx::Database,
+    sqlx::types::Json<proto::PartitionTemplate>: sqlx::Decode<'q, DB>,
+{
+    fn decode(
+        value: <DB as sqlx::database::HasValueRef<'q>>::ValueRef,
+    ) -> Result<Self, Box<dyn std::error::Error + 'static + Send + Sync>> {
+        Ok(Self(
+            <sqlx::types::Json<proto::PartitionTemplate> as sqlx::Decode<'_, DB>>::decode(value)?
+                .0
+                .into(),
+        ))
+    }
+}
+
+/// The default partitioning scheme is by each day according to the "time" column.
+pub static PARTITION_BY_DAY_PROTO: Lazy<Arc<proto::PartitionTemplate>> = Lazy::new(|| {
+    Arc::new(proto::PartitionTemplate {
+        parts: vec![proto::TemplatePart {
+            part: Some(proto::template_part::Part::TimeFormat(
+                "%Y-%m-%d".to_owned(),
+            )),
+        }],
+    })
+});
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use sqlx::Encode;
+
```

Comment:
Lets add a fixture test to ensure the default derived template isn't changed, and add a note why:

```suggestion

    /// This test asserts the default derived partitioning scheme with no
    /// overrides.
    ///
    /// Changing this default during the lifetime of a cluster will cause the
    /// implicit (not overridden) partition schemes to change, potentially
    /// breaking the system invariant that a given primary keys maps to a
    /// single partition.
    ///
    /// You shouldn't be changing this!
    #[test]
    fn test_default_template_fixture() {
        let ns = NamespacePartitionTemplateOverride::default();
        let table = TablePartitionTemplateOverride::new(None, &ns);
        let got = table.parts().collect::<Vec<_>>();

        assert_matches!(got.as_slice(), [TemplatePart::TimeFormat("%Y-%m-%d")]);
    }

```

Reply:
Thank you for writing up the reasoning! I've added this in [`0a6944a` (#7790)](https://github.com/influxdata/influxdb_iox/pull/7790/commits/0a6944a12f3e1492aa1083ab5862d32a69aa7bc7)